### ISSUE TITLE
Sorting scales poorly with many plugins or userlist rules

### ISSUE THREAD

**Issue by MacSplody** [loot/loot#1370]
Originally brought up in Vortex Support
[Disabled Plugins, Staging Folders, and Excessive Sort Times](https://forums.nexusmods.com/index.php?/topic/8977018-disabled-plugins-staging-folders-and-excessive-sort-times/)

Information supplied by @nt5raham and @Pickysaurus
> Basically, in this scenario. The user has 1,200 plugins. ~500 of which are disabled. Sorting them alone takes 7 minutes but including the userlist.yaml extends this time up to 61 minutes.
> With the same user list but deleting all disabled plugins the time cuts down to 5 minutes

[userlist.zip](https://github.com/loot/loot/files/5540731/userlist.zip)

[LOOTDebugLog - with Userlist.zip](https://github.com/loot/loot/files/5540733/LOOTDebugLog.-.with.Userlist.zip)

[LOOTDebugLog - no Userlist.zip](https://github.com/loot/loot/files/5540735/LOOTDebugLog.-.no.Userlist.zip)

**Comment by Ortham** [comment]
So without a userlist, the timings are:

```
line 263276: [11:41:37.853698] [info]: Current load order: 
line 264564: [11:41:37.954729] [trace]: Adding Master Flag edge from "3DNPC.esp" to "018Auri.esp".
+60k lines, +8 seconds
line 323875: [11:41:45.021463] [trace]: Adding Hardcoded edge from "Skyrim.esm" to "Expressive Facegen Morphs.esl".
line 324010: [11:41:45.753082] [trace]: Skipping group edge from "Audio Overhaul Skyrim.esp" to "3DNPC.esp" as it would create a cycle.
+28k lines, +11 seconds
line 352612: [11:41:56.495462] [trace]: Adding Overlap edge from "018Auri.esp" to "3DNPC Alternative Locations.esp".
+10k lines, +12 seconds
line 362547: [11:42:08.893214] [trace]: Adding Tie Break edge from "12FemaleBrows.esp" to "018Auri.esp".
+11k lines, +6 min, 33 seconds
line 373683: [11:48:41.002712] [info]: Calculated order: 

[11:48:41.189757] [trace]: Checking if plugin "dz_moonstone.esp" is active.
[11:49:33.311399] [trace]: Checking if plugin "DynDOLOD.esp" is active.
[11:49:37.320695] [trace]: User has accepted sorted load order, applying it.
[11:49:37.320975] [trace]: Getting load order.
[11:49:37.419197] [info]: Setting load order.
[11:49:37.551123] [info]: Load order set successfully.
```

The very poor scaling doesn't surprise me, and nor does that it's the tie break edges that take the most time to process. 

It's interesting that generating the derived metadata for display after sorting takes 52 seconds, it looks like there aren't any big jumps in time, suggesting that there's no one thing causing the slowness there. However, plugins are checked to see if they're active many times over, suggesting that the process is doing a lot of unnecessary work.

---

With a userlist, the timings are:

```
line 263305: [10:47:31.641203] [info]: Current load order: 
line 264593: [10:47:31.725946] [trace]: Adding Master Flag edge from "3DNPC.esp" to "018Auri.esp".
+60k lines, +7 seconds
line 323937: [10:47:38.828828] [trace]: Adding Hardcoded edge from "Skyrim.esm" to "Expressive Facegen Morphs.esl".
line 324072: [10:47:39.513318] [trace]: Skipping group edge from "BetterDesc IllumAtro BreezeBase ComPatch.esp" to "018Auri.esp" as it would create a cycle.
+650k lines, 3 min, 41 seconds
line 973763: [10:51:20.540860] [trace]: Adding Overlap edge from "Ambriel.esp" to "018Auri.esp".
+5k lines, 50 seconds
line 979014: [10:52:10.776974] [trace]: Adding Tie Break edge from "018Auri.esp" to "CaesiaFollower.esp".
+11k lines, +30 min
line 989764: [11:32:08.365460] [info]: Calculated order: 

line 991080: [11:32:08.743000] [trace]: Checking if plugin "UHDAP - MusicHQ.esp" is active.
line 1494157: [11:32:58.880800] [trace]: Checking if plugin "Atlas Map Markers ComPatch.esp" is active.
```

The increase in time taken on group edges makes sense considering the userlist adds ~ 1200 group definitions. It's interesting that this then causes the tie break edges to take so much longer - I suspect this is because the graph ends up with more edges between plugins before it hits the tie-breaking stage, and so there are a lot of paths to follow when checking if adding an edge would cause a cycle.

As expected, generating the derived metadata still takes ~50s because while a userlist may mean there's more data to display, deriving from it doesn't involve a complexity explosion.

However, I'm not seeing that 61 minute sort time in the logs, it's more like 45 min.

---

In both logs, there are 1287 plugins - that removing 500 plugins would drop the sorting time down to 5 min sounds about right.

The ~50 seconds spent generating derived metadata seems like something that could be sped up relatively easily, but it's also not that significant next to the 7 min sort without a userlist and 45 min sort with a userlist.

The trouble with speeding up tie-break edges is that adding an edge can affect whether or not the next edge would cause a cycle, so the possible combinations of vertices are evaluated in series. I think it's probably the check for a cycle that is the slowest part of each iteration, but that assumption needs testing.

I don't think I've ever been so glad that my mouse scroll wheel has a free-scrolling mode...

**Comment by Pickysaurus** [comment]
Thanks for the update Ortham.

I can say that the user list sort did take an hour from clicking "Sort now" to the app finishing. I also did the sort in Vortex, which gave a ~3660 sort time measurement in the log. You can see this in the Vortex.log file I sent to MacSplody. 

Perhaps for load order of this extreme length, an option to only sort active plugins would be prudent? Or would removing excess/unused entires from the userlist.yaml help?

**Comment by nt5raham** [comment]
Thanks to all of you for looking into this!

As a test, I removed the group assignment for 507 disabled plugins.  Sort time was 3:38 in Vortex.

In my situation, the disabled plugins have been merged using zMerge, and the mods need to remain enabled so the merged plugins can access the resources.  The group assignments are used to block the plugins into a contiguous load order when creating/changing/updating the merge.  The sort order of the disabled plugins (and thus the groups) have no purpose for game play as far as I know.

**Comment by Ortham** [comment]
> Perhaps for load order of this extreme length, an option to only sort active plugins would be prudent? Or would removing excess/unused entires from the userlist.yaml help?

That's an idea, but I'd rather solve the underlying issue of the algorithm scaling incredibly poorly, or someone else will come along with not that many more plugins and hit the issue again.

I spent some time at the weekend thinking about this, I've got a few ideas but no progress to report just yet.

**Comment by Pickysaurus** [comment]
Thanks for keeping us updated :)

**Comment by nt5raham** [comment]
Hello all, just checking in to see if there's been any progress, or if I can be of some assistance.  Thanks!

**Comment by MacSplody** [comment]
Another set of tests done by ra2phoenix
[Discord link](https://discord.com/channels/473542112974077963/496196499890241546/852425395634700298)
1023 plugins
No disabled plugins
Large number of groups in userlist.

[userlist.zip](https://github.com/loot/loot/files/6629015/userlist.zip)

[LOOTDebugLog.zip](https://github.com/loot/loot/files/6628847/LOOTDebugLog.zip)

**Comment by MacSplody** [comment]
![image](https://user-images.githubusercontent.com/1944639/122160786-70d72280-ce68-11eb-82be-70f256ee513d.png)

It may be worth testing with increasing increments of plugins to see if it's a gradual slowdown or a sudden spike.

**Comment by Sorbe** [comment]
An uninformed observation- if LOOT Is actually "sorting each group", that suggests perf testing something like "mergesort".
Perhaps even multi-threaded parallel mergesort given the number of available cores in almost all modern PC CPUs.

https://en.wikipedia.org/wiki/Merge_sort

**Comment by MacSplody** [comment]
> @MacSplody, what did @Xanza do with the numerous plugins after the merge?
> 

They removed them from the data folder through their mod manager.

**Comment by nt5raham** [comment]
@MacSplody As an alternative solution, would it be feasible for LOOT to exclude Ghosted plugins from the sort process?

Ghosted plugins:

- Are supported by all of the major mod managers and LOOT.
- Already solve the problem of exceeding the max file handle limit, another problem associated with having too many plugins in \Data.

This could make Ghosted plugins a two-in-one solution.

**Comment by Pickysaurus** [comment]
> > > @MacSplody, what did @Xanza do with the numerous plugins after the merge?
> > 
> > 
> > @nt5raham They removed them from the data folder through their mod manager.
> 
> @Pickysaurus What has been a large stumbling block for me is that Vortex does not "support" removing the plugins from the data folder for Enabled mods. A topic for a Vortex thread.

Ghosting should be a solution here, it's just that LOOT doesn't ignore these ghosted plugins properly.

**Comment by Ortham** [comment]
> Ghosting should be a solution here, it's just that LOOT doesn't ignore these ghosted plugins properly.

LOOT intentionally doesn't ignore ghosted plugins, because they can be merged into the Bashed Patch, and when that happens any conflicts are resolved according to the plugins' load order.

**Comment by Pickysaurus** [comment]
I don't really understand why other tools are expected to use Wyrebash practices. Maybe sorting ghosts could be an optional parameter for the sake of performance in large load orders?

**Comment by Ortham** [comment]
> I don't really understand why other tools are expected to use Wyrebash practices.

Well, IIRC Wrye Bash was the first tool to introduce ghosting plugins, and compatibility is nice. Ghosting wasn't invented to solve LOOT's performance problems, and I don't think using it in that way is a good fit - almost, but not quite.

**Comment by Pickysaurus** [comment]
I mean, we could just introduce a new standardised state for plugins that LOOT would ignore? Something like `Skyrim.esp.nosort`?

**Comment by Ortham** [comment]
Off the top of my head, I think LOOT will ignore anything that's not .esp, .esm, .esl or a .ghost suffix on any of those three. You could probably try adding whatever other extension you want and see if that works for you, before proposing it to Vortex and/or other mod managers.

**Comment by Alaxouche-SE** [comment]
Hey; 
I hope this might help. I have this kind of issue with my modlist, so i ran LOOT with debug logger option. Here is the result:
Debug Log: [https://drive.google.com/file/d/1Sns1ROG4HOfT-R4NFopxK_Ka8JPF-D5j/view?usp=sharing](url)
Number of Plugins: 1052
Time to sort: Between 5 to 10 minutes
CPU: I7-7700
Storage: SSD 512go
Also i run LOOT as executable (the real app, not the one already in Mo2) through Mo2 v 2.4.4
If you need more infos tell me!

**Comment by LyraKeaton** [comment]
Hello there.
I was told I could post my log here for potential use. 
Here is my Debug Log: [https://drive.google.com/file/d/1LW1Iab2JY_gETiK1Jf3-4EgxEqO7tful/view?usp=sharing](url)
Number of Plugins: 2218
Time to sort: 1 hour and 52 minutes 
CPU: AMD Ryzen 5 2600X
Storage: HDD Seagate (5 Terabytes)
I ran LOOT (v0.18.4) as an executable through Mod Organizer 2 (v2.4.4)

**Comment by sibir-ine** [comment]
Can you please link the file directly instead of through Google Drive? You can just drag it into the message while editing it.

**Comment by LyraKeaton** [comment]
Updated with compressed zip file of the log, as Github won't accept files over 25MB.

**Comment by Ortham** [comment]
I was able to compress the log to 22.5 MB using 7z's PPMd compression method, so I've uploaded @LyraKeaton's log directly in this issue and updated the link.

**Comment by diskdevl** [comment]
Hopefully this info helps a bit. 

Number of Plugins: 2154
Time to sort: 3 hour and 19 minutes
CPU: AMD Ryzen 7 3800X 3.9Ghz
Storage: Samsung 860 EVO SSD (1 TB)
I ran LOOT (v0.18.4) as an executable through Mod Organizer 2 (v2.5.0alpha-3

Sorry for having to split the file, it compressed down to 31 MB (25 MB limit here apparently). Just delete the ".deletethisextension.zip" part on the second file and you should be able to rejoin them. 

[LOOTDebugLog.7z.001.zip](https://github.com/loot/loot/files/9689814/LOOTDebugLog.7z.001.zip)
)
[LOOTDebugLog.7z.002.deletethisextension.zip](https://github.com/loot/loot/files/9689817/LOOTDebugLog.7z.002.deletethisextension.zip)

**Comment by sibir-ine** [comment]
I managed to further compress it down to 23 MB by using a word size of 12 (& compressed it as a ZIP for convenience).

**Comment by Ortham** [comment]
I made some changes to libloot today that reduced the time to sort a 1619 plugin load order from 17 min 10 secs to 5 min 58 secs (almost 1/3rd the time). 

I'd be interested in other peoples' results: you can try the changes out by downloading the 32-bit or 64-bit libloot (matching whether you're using 32-bit or 64-bit LOOT) from [here](https://loot.jfrog.io/ui/native/libloot/sorting-performance/) and replacing the `loot.dll` in your LOOT v0.18 install with the one in the downloaded archive.

**Comment by sibir-ine** [comment]
In [Discord](https://discord.com/channels/473542112974077963/496196499890241546/1049076433537994792), diskdevl pointed out a few logging improvements that would help in testing this. One would be explicitly logging when sorting ends. The other was logging when a sort is started & ends when debug mode isn't enabled as well, but I imagine we actually want full debug logs here to see what else might still be taking a lot of time. Still, I imagine it couldn't hurt to implement anyway.

**Comment by diskdevl** [comment]
To badly paraphrase Star Wars, "These are numbers I haven't seen in a long time. A looooooong time. " =)

Number of Plugins: 2,303
Time to sort: 38 minutes !!!
CPU: AMD Ryzen 7 3800X 3.9Ghz
Storage: Samsung 970 EVO Plus NVM M.2
I ran LOOT (v0.18.6) as an executable through Mod Organizer 2 (v2.5.0alpha-6)

There were probably only a handful (<5) of plugins that required moving around in the load order this time but that's still beyond impressive. Nice job! Did it change to an incremental sort of some kind instead of resorting everything ? 

[LOOTDebugLog_120522_7zip ultra PPMd 256 dic word 12.7z.zip](https://github.com/loot/loot/files/10158114/LOOTDebugLog_120522_7zip.ultra.PPMd.256.dic.word.12.7z.zip)

**Comment by sibir-ine** [comment]
For what it's worth, with 527 plugins, sorting went from 33 seconds to 8 seconds.

**Comment by Ortham** [comment]
> For what it's worth, with 527 plugins, sorting went from 33 seconds to 8 seconds.

Is that 33 seconds with stock LOOT v0.18.6, or with the optimised build I linked to [a few posts above](https://github.com/loot/loot/issues/1370#issuecomment-1335616793)?

Can you share your debug log?

**Comment by diskdevl** [comment]
It’ll probably be Monday before I have a chance to, but after that sure.

**Comment by Ortham** [comment]
I've just optimised the new tie-breaking logic some more: it now sorts the 1619 plugin load order that I'm testing with in 55 seconds. That's more than 3 times faster than before, and more than 18 times faster than before I started optimising last week.

The download link is [here](https://loot.jfrog.io/ui/native/libloot/new-tie-break-logic/) (it's the same link as above).

Note that this new tie-break logic does give different results from the previous logic, but it shouldn't matter because a tie only exists when LOOT thinks it doesn't matter which order a pair of plugins loads in. The new logic should also be faster the closer the sorted load order is to the old load order (i.e. the fewer necessary changes there are, the less work it does breaking ties).

With the latest optimisations, tie-breaking is no longer the slowest part of sorting, it's pretty similar to handling groups and overlaps:

```
[19:50:33.797227] [info]: Current load order: 
[19:50:33.821085] [trace]: Adding Master Flag edge from "Skyrim.esm" to "018Auri.esp".
+3.8 seconds
[19:50:37.591542] [trace]: Adding Hardcoded edge from "Skyrim.esm" to "ccFFBSSE001-ImperialDragon.esl".
[19:50:37.649061] [trace]: Skipping group edge from "Immersive Wenches.esp" to "Skyrim.esm" as it would create a cycle.
+17.0s
[19:50:54.603525] [trace]: Skipping vertex for "Skyrim.esm": the plugin contains no override records.
+17.5s
[19:51:12.109742] [debug]: The plugin "Skyrim.esm" loads at the end of the new load order so far.
+16.4
[19:51:28.453364] [info]: Calculated order: 

= 54.7
```

Hopefully those of you with huge load orders will see similar benefits (and nothing going wrong).

**Comment by digitaljc** [comment]
Working for me in Vortex; literal night and day difference; minutes to <1m on open w/ 1176 plugins enabled. 
Also, deployment process massively reduced. I do have to put up with the 'Vortex Installation Corrupted' message. Select 'Quit' to continue and viola, in business! Thank you!!!

**Comment by Ortham** [comment]
Another update: I've made some improvements that speed up adding group edges. When using the huge userlist linked to in the OP, the time it takes to add group edges has dropped from 332 seconds to 34 seconds.

Without a userlist the gains are more modest but still significant. With my 1619 plugin load order the time taken to add group edges has dropped from 17s to 1.8s.

These latest changes are now available at the same link as I last posted above.

**Comment by Pickysaurus** [comment]
Sounds like you guys are making excellent progress, thanks for looking at this! <3

**Comment by diskdevl** [comment]
LOOT has gone from 3.5 hours per run, to 38 minutes with libloot-0.18.2-16-g2bd3682, and now to just over FOUR MINUTES using libloot-0.18.2-24-g115a608. I think my head is going to explode, wow. 

Number of Plugins: 2,313
Time to sort: 00:04:24 
CPU: AMD Ryzen 7 3800X 3.9Ghz
Storage: Samsung 970 EVO Plus NVM M.2
I ran LOOT (v0.18.6) as an executable through Mod Organizer 2 (v2.5.0alpha-6)

My apologies for the file split, I tried about a dozen combinations and just could not get 7zip to bring it below the 25 MB limit - closest I managed was 27.6 MB. If you find settings that work please let me know so I can use them next time.
[LOOTDebugLog_122222_split.7z.001.zip](https://github.com/loot/loot/files/10209987/LOOTDebugLog_122222_split.7z.001.zip)
[LOOTDebugLog_122222_split.7z.002.zip](https://github.com/loot/loot/files/10209992/LOOTDebugLog_122222_split.7z.002.zip)

**Comment by Ortham** [comment]
@diskdevl Thanks for that. It looks like the actual sorting took 66 seconds:

```
[07:47:16.899012] [info]: Beginning sorting operation.
[07:47:16.899067] [info]: Loading the current load order state.
[07:47:18.156990] [trace]: Getting active plugins.
...
[07:47:18.754907] [trace]: Starting plugin loading.
...
[07:47:19.603209] [trace]: Evaluating condition: checksum("Ivarstead.esp", 095A4B0C)
...
[07:47:24.966609] [trace]: Masterlist group "DLC" directly loads after groups ""
...
[07:47:25.240579] [info]: Current load order: 
...
[07:47:25.257004] [trace]: Adding Master Flag edge from "Skyrim.esm" to "TerrainLodRedone.esp".
+7.3s
[07:47:32.527840] [trace]: Adding Hardcoded edge from "Skyrim.esm" to "DyeManagerDefaults.esl".
+0.2s
[07:47:32.703628] [trace]: Skipping group edge from "ccrmssse001-necrohouse.esl" to "Skyrim.esm" as it would create a cycle.
+7.6
[07:47:40.311671] [trace]: Skipping vertex for "Skyrim.esm": the plugin contains no override records.
+41.4s
[07:48:21.731317] [trace]: Adding Tie Break edge from "Skyrim Project Optimization - No Homes - Full Version.esm" to "Vanilla Script MicroOptimizations.esl".
+9.2s
[07:48:30.891480] [trace]: Checking plugin graph for cycles...
[07:48:30.908455] [trace]: Performing topological sort on plugin graph...
[07:48:30.909173] [trace]: Checking uniqueness of path through plugin graph...
[07:48:30.909785] [info]: Calculated order: 
...
[07:49:58.784143] [trace]: Checking if plugin "Dragonborn.esm" is active.
[07:50:05.936330] [trace]: User has accepted sorted load order, applying it.
[07:50:05.936394] [trace]: Getting load order.
[07:50:05.953658] [info]: Setting load order.
...
[07:50:05.982426] [info]: Load order set successfully.
```

With about 10 seconds for loading plugins and then 88 seconds of mostly checking if plugins are active (which seems excessive, I'll look into that).

**Comment by diskdevl** [comment]
Sorry, I must've been accidentally looking at when I hit apply. 66 seconds is even better. =)

**Comment by Ortham** [comment]
All of the optimisations other than the tie-breaking optimisations are in the newly-released libloot v0.18.3 (@Pickysaurus you may be interested in that), which is now uses by the latest LOOT master branch build - I've also:

* increased LOOT's default logging verbosity to include info messages and adjusted log messages so that's not too verbose
* added some more log messages to clearly indicate the end of sorting (and at trace verbosity the start of each type of edge processing)
* optimised some of the load order / active plugin checks that happen when loading plugin data, which may noticeably speed up the post-sorting UI refresh (it'll at least cause less log spam).

With libloot v0.18.3 I see a 5x performance improvement, and I plan to release a new version of LOOT v0.18 with those changes soon (probably before Christmas). 

I've decided to hold the tie-breaking optimisations back for LOOT v0.19.0 as they may change the result sorting gives, and that could cause plugins that are unrelated but that load BSAs that overwrite the same resources to change their relative positions and so still affect things in-game. Putting such a change in a patch release therefore doesn't seem like the right thing to do. My plan is to release libloot and LOOT v0.19.0 soon after the 10th of January. 

For now, users can continue to drop the `new-tie-break-logic` branch build of libloot into their LOOT v0.18 installs to get the tie-breaking optimisations, as that branch is still binary-compatible with LOOT v0.18.

**Comment by Ortham** [comment]
Cloud and Proud sorted their load order of 2803 plugins with LOOT v0.18.6 and it took them around 8 hours.

They're going to re-run sorting using the `new-tie-break-logic` branch of libloot, and I'll update this comment with the results once they've been shared.

**Comment by Ortham** [comment]
The latest `differentiate-group-edge-types` LOOT build now includes a libloot update that splits the plugin graph into two: one to sort master-flagged plugins and the other to sort the others.

I've seen a ~25% speed improvement because of this, for two reasons:

1. most edges added are between master-flagged and non-master-flagged plugins, and removing those edges speeds up checking if adding others would cause cycles because the graph is simpler to navigate
2. parts of sorting like adding overlap edges are roughly `O(N^2)` for `N` plugins, so splitting the graph into two smaller graphs gives better performance.

The change does make a couple of behavioural changes (data/metadata that contradict master flags will be ignored instead of causing cycle errors, and tie breaking may produce different results), but I don't think they're major.

@diskdevl could you try sorting [this LOOT build](https://loot.jfrog.io/ui/native/loot/differentiate-group-edge-types/) and share the log it produces?

<details>
<summary>
Here's the relevant log entries:
</summary>

```
[15:31:36.960806] [info]: Beginning sorting operation.
[15:31:36.960871] [trace]: Loading the current load order state.
[15:31:37.646737] [trace]: Getting active plugins.
[15:31:37.646833] [trace]: Getting load order.
[15:31:37.825265] [trace]: Starting plugin loading.
+1.36s
...
[15:31:39.192235] [trace]: Evaluating condition: checksum("018Auri.esp", 165E7B6D)
+0.93s
...
[15:31:40.118705] [debug]: Current load order:
+0.03s
...

START OF MASTER-FLAGGED-PLUGIN SORTING

[15:31:40.148178] [trace]: Adding edges based on plugin data and non-group metadata...
+0.004s
...
[15:31:40.152202] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
+0.01s
...
[15:31:40.162299] [trace]: Adding edges based on plugin group memberships...
+0.006s
...
[15:31:40.168730] [trace]: Adding edges for overlapping plugins...
+0.01s
...
[15:31:40.178039] [trace]: Adding edges to break ties between plugins...
+0.003s
...
[15:31:40.181215] [trace]: Checking plugin graph for cycles...
[15:31:40.181353] [trace]: Performing topological sort on plugin graph...
[15:31:40.181377] [trace]: Checking uniqueness of path through plugin graph...

START OF NON-MASTER-FLAGGED-PLUGIN SORTING

[15:31:40.181407] [trace]: Adding edges based on plugin data and non-group metadata...
+4.2s
...
[15:31:44.379040] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
+0.05s
...
[15:31:44.425380] [trace]: Adding edges based on plugin group memberships...
+0.85s
...
[15:31:45.284051] [trace]: Adding edges for overlapping plugins...
+11.24s
...
[15:31:56.515930] [trace]: Adding edges to break ties between plugins...
+13.86s
...
[15:32:10.382422] [trace]: Checking plugin graph for cycles...
[15:32:10.392849] [trace]: Performing topological sort on plugin graph...
[15:32:10.393066] [trace]: Checking uniqueness of path through plugin graph...
[15:32:10.393447] [debug]: Calculated order:
+0.11s
...
[15:32:10.501908] [trace]: Checking if plugin "Skyrim.esm" is active.
+1.2s
...
[15:32:11.698414] [trace]: Checking if plugin "arnima.esm" is active.
[15:32:14.790516] [trace]: User has rejected sorted load order, discarding it.
```

</details>

**Comment by diskdevl** [comment]
Sure, due to the holiday it will probably be Monday before I have time though, sorry.

Thank you for the continuing great updates !

**Comment by Ortham** [comment]
Rayden on Discord tested with a 1950 plugin load order. With the latest `differentiate-group-edge-types` LOOT build sorting took just under 27 seconds.

Debug logs:

[LOOTDebugLog.cd758f74.txt.zip](https://github.com/loot/loot/files/10332602/LOOTDebugLog.cd758f74.txt.zip) (latest build)

I'll upload the v0.18.6 log once it's been shared.

<details>
<summary>Relevant log entries</summary>

```
[16:51:11.741184] [info]: Beginning sorting operation.
[16:51:11.741224] [trace]: Loading the current load order state.
[16:51:12.450237] [trace]: Getting active plugins.
[16:51:12.451063] [trace]: Getting load order.
[16:51:12.668819] [trace]: Starting plugin loading.
...
[16:51:13.097729] [trace]: Evaluating condition: active("Complete Crafting Overhaul_Remastered.esp") and ((active("Cloaks.esp") and not active("1nivWICCloaks_COS_CCOR_Patch.esp")) or (not active("Cloaks.esp") and not active("1nivWICCloaks_CCOR_Patch.esp")))
...
[16:51:16.758749] [debug]: Current load order:
...
[16:51:16.774464] [trace]: Adding edges based on plugin data and non-group metadata...
...
[16:51:16.775885] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
...
[16:51:16.777763] [trace]: Adding edges based on plugin group memberships...
...
[16:51:16.778753] [trace]: Adding edges for overlapping plugins...
...
[16:51:16.781317] [trace]: Adding edges to break ties between plugins...
...
[16:51:16.782028] [trace]: Checking plugin graph for cycles...
[16:51:16.782086] [trace]: Performing topological sort on plugin graph...
[16:51:16.782099] [trace]: Checking uniqueness of path through plugin graph...
[16:51:16.782114] [trace]: Adding edges based on plugin data and non-group metadata...
+6.39s
...
[16:51:23.171791] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
+0.04s
...
[16:51:23.207334] [trace]: Adding edges based on plugin group memberships...
+0.5s
...
[16:51:23.709263] [trace]: Adding edges for overlapping plugins...
+18.33s
...
[16:51:42.041442] [trace]: Adding edges to break ties between plugins...
+1.49s
...
[16:51:43.534185] [trace]: Checking plugin graph for cycles...
[16:51:43.544224] [trace]: Performing topological sort on plugin graph...
[16:51:43.544524] [trace]: Checking uniqueness of path through plugin graph...
[16:51:43.544887] [debug]: Calculated order:
+0.14s
...
[16:51:43.681024] [trace]: Checking if plugin "Skyrim.esm" is active.
+0.34s
...
[16:51:44.015105] [trace]: Checking if plugin "Northern Roads - Rocks Patch.esp" is active.
[16:51:47.352578] [trace]: User has accepted sorted load order, applying it.
```

Edge counts:

* 0 master flag
* 2589 master
* 7 masterlist requirement
* 69 masterlist load after
* 0 user requirement
* 392 user load after
* 85 hardcoded
* 43748 masterlist group
* 0 user group
* 83247 record overlap
* 8 asset overlap
* 1257 tie breaks

Total: 131,402

</details>

**Comment by diskdevl** [comment]
As requested. 

2,309 active plugins currently. Using the new 'differentiate-group-edge-types', new log file attached. 

[15:43:29.666318] [info]: Beginning sorting operation.
...
[15:44:17.410142] [trace]: Checking if plugin "Occlusion.esp" is active.
[15:44:25.282877] [trace]: User has accepted sorted load order, applying it.
...
[15:44:25.324836] [debug]: Load order set successfully.


[LOOTDebugLog_010223_diff-group-edge-types.7z.zip](https://github.com/loot/loot/files/10333805/LOOTDebugLog_010223_diff-group-edge-types.7z.zip)

**Comment by Ortham** [comment]
Thanks, looks like sorting took 40.49s and the post-sorting UI data loading took 0.53s, a nice improvement from the previous 66s and 88s respectively. The breakdown for non-master-flagged plugin sorting is:

```
[15:43:36.396781] [trace]: Adding edges based on plugin data and non-group metadata...
+9.49s
...
[15:43:45.892348] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
+0.05s
...
[15:43:45.948828] [trace]: Adding edges based on plugin group memberships...
+0.9s
...
[15:43:46.863044] [trace]: Adding edges for overlapping plugins...
+26.93s
...
[15:44:13.791535] [trace]: Adding edges to break ties between plugins...
+2.87s
...
[15:44:16.659260] [debug]: Calculated order:
```

**Comment by Ortham** [comment]
For any future readers: I've renamed the `differentiate-group-edge-types` branch to `update-libloot` as the old name hasn't really fit for a while.

**Comment by Ortham** [comment]
Changed merged as of edbbc8b730bf0ea8974c10012db14639a94be1eb, closing this.

**Comment by Ortham** [comment]
Thanks to everyone who contributed logs and took the time to test changes and provide feedback. As a final example of the progress made, I ran sorting with v0.18.6 and a build of commit 7f011997007a9e0d11ed10806de22b1b5be75705 and different numbers of plugins,

| LOOT build | Number of plugins | Time / s |
|-------------|--------------------|---------|
| 0.18.6 | 450 | 10.0
| 0.18.6 | 924 | 142
| 0.18.6 | 1619 | 1186
| 7f011997007a9e0d11ed10806de22b1b5be75705 | 450 | 1.41
| 7f011997007a9e0d11ed10806de22b1b5be75705 | 1619 | 30.7
| 7f011997007a9e0d11ed10806de22b1b5be75705 | 3228 | 233
| 7f011997007a9e0d11ed10806de22b1b5be75705 | 4620 | 674

So sorting is now faster and scales better: the numbers above suggest the scaling with v0.18.6 is ~ `O(N^3.73)` and sorting with 7f011997007a9e0d11ed10806de22b1b5be75705  is ~ `O(N^2.93)`. If I'd tried sorting 4620 plugins with 0.18.6 it could have taken me over 16 hours!

<details>
<summary>Log extracts</summary>

0.18.6, 450 plugins, took 10s

```
[14:24:51.467207] [info]: Current load order: 
[14:25:01.439646] [info]: Calculated order: 
```

0.18.6, 924 plugins, took 142s

```
[14:07:36.952255] [info]: Current load order: 
[14:09:58.940153] [info]: Calculated order: 
```

0.18.6, 1619 plugins, took 1186s:

```
[10:23:59.001236] [info]: Current load order: 
[10:23:59.022501] [trace]: Adding Master Flag edge from "3DNPC.esp" to "018Auri.esp".
[10:24:30.246867] [trace]: Adding Overlap edge from "018Auri.esp" to "3DNPC Alternative Locations.esp".
[10:24:04.626322] [trace]: Skipping group edge from "Gildergreen Regrown.esp" to "3DNPC.esp" as it would create a cycle.
[10:24:30.246867] [trace]: Adding Overlap edge from "018Auri.esp" to "3DNPC Alternative Locations.esp".
[10:25:11.913341] [trace]: Adding Tie Break edge from "018Auri.esp" to "0Abigail.esp".
[10:43:45.329721] [info]: Calculated order: 
```

7f011997007a9e0d11ed10806de22b1b5be75705, 450 plugins, took 1.41s

```
[14:45:01.424951] [debug]: Current load order:
[14:45:02.831867] [debug]: Calculated order:
```

7f011997007a9e0d11ed10806de22b1b5be75705, 1619 plugins, took 30.7s

```
[13:27:48.581807] [debug]: Current load order:
[13:27:48.606238] [trace]: Adding edges based on plugin data and non-group metadata...
[13:27:48.610121] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
[13:27:48.612158] [trace]: Adding edges based on plugin group memberships...
[13:27:48.619768] [trace]: Adding edges for overlapping plugins...
[13:27:48.629226] [trace]: Adding edges to break ties between plugins...
[13:27:48.632527] [trace]: Adding edges based on plugin data and non-group metadata...
[13:27:52.834960] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
[13:27:52.869281] [trace]: Adding edges based on plugin group memberships...
[13:27:53.705758] [trace]: Adding edges for overlapping plugins...
[13:28:04.931752] [trace]: Adding edges to break ties between plugins...
[13:28:19.229145] [debug]: Calculated order:
```

7f011997007a9e0d11ed10806de22b1b5be75705, 3228 plugins, took 233s

```
[12:36:37.721688] [debug]: Current load order:
[12:36:37.771162] [trace]: Adding edges based on plugin data and non-group metadata...
[12:36:37.786982] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
[12:36:37.791121] [trace]: Adding edges based on plugin group memberships...
[12:36:37.812003] [trace]: Adding edges for overlapping plugins...
[12:36:37.921221] [trace]: Adding edges to break ties between plugins...
[12:36:37.933529] [trace]: Adding edges based on plugin data and non-group metadata...
[12:36:56.225983] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
[12:36:56.296716] [trace]: Adding edges based on plugin group memberships...
[12:36:58.250828] [trace]: Adding edges for overlapping plugins...
[12:38:46.911159] [trace]: Adding edges to break ties between plugins...
[12:40:30.673612] [debug]: Calculated order:
7f011997007a9e0d11ed10806de22b1b5be75705, 3228 plugins:
```

7f011997007a9e0d11ed10806de22b1b5be75705, 4620 plugins, took 674s

```
[13:39:07.220846] [debug]: Current load order:
[13:39:07.287994] [trace]: Adding edges based on plugin data and non-group metadata...
[13:39:07.298949] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
[13:39:07.304587] [trace]: Adding edges based on plugin group memberships...
[13:39:07.348306] [trace]: Adding edges for overlapping plugins...
[13:39:07.596079] [trace]: Adding edges to break ties between plugins...
[13:39:07.625965] [trace]: Adding edges based on plugin data and non-group metadata...
[13:39:48.136750] [trace]: Adding edges for implicitly active plugins and plugins with hardcoded positions...
[13:39:48.237529] [trace]: Adding edges based on plugin group memberships...
[13:39:51.337214] [trace]: Adding edges for overlapping plugins...
[13:44:54.856614] [trace]: Adding edges to break ties between plugins...
[13:50:21.755729] [debug]: Calculated order:
```
